flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java

```diff
@@ -729,13 +738,33 @@ private void registerTableSinkInternal(String name, TableSink<?> tableSink) {
 	}
 
 	private void createCatalogFunction(CreateFunctionOperation createFunctionOperation) {
-		Catalog catalog = getCatalogOrThrowException(createFunctionOperation.getFunctionIdentifier().getCatalogName());
 		String exMsg = getDDLOpExecuteErrorMsg(createFunctionOperation.asSummaryString());
 		try {
-			catalog.createFunction(
-				createFunctionOperation.getFunctionIdentifier().toObjectPath(),
-				createFunctionOperation.getCatalogFunction(),
-				createFunctionOperation.isIgnoreIfExists());
+			CatalogFunction function = createFunctionOperation.getCatalogFunction();
+			if (function.isTemporary()) {
+				boolean exist = functionCatalog.hasTemporaryCatalogFunction(
+					createFunctionOperation.getFunctionIdentifier());
+				if (!exist) {
+					FunctionDefinition functionDefinition = FunctionDefinitionUtil.createFunctionDefinition(
+						createFunctionOperation.getFunctionName(),
+						function);
+					registerFunctionInFunctionCatalog(
+						createFunctionOperation.getFunctionIdentifier(),
+						functionDefinition);
+				} else if (!createFunctionOperation.isIgnoreIfExists()) {
+					throw new FunctionAlreadyExistException(FunctionCatalog.TEMPORARY_CATALOG,
```

Comment:
just throw a ValidationException directly? feels a bit weird to throw an exception and catch it right after in the same method

Reply:
Agree. I will change to Validation Exception here.